nixos/lib/make-zfs-image.nix

```diff
@@ -0,0 +1,340 @@
+{ lib
+, pkgs
+, # The NixOS configuration to be installed onto the disk image.
+  config
+
+, # The size of the root disk, in megabytes.
+  rootSize ? 2048
+
+, # The name of the ZFS pool
+  bootPoolName ? "boot"
+
+, # size of the boot disk
+  bootSize ? 1024
+, # features on the boot pool
+  # Note: the pool has no features enabled by default,
+  # and only this list of features is enabled. This list
+  # must match GRUB's supported list.
+  # See: https://github.com/openzfs/openzfs-docs/blame/master/docs/Getting%20Started/Debian/Debian%20Stretch%20Root%20on%20ZFS.rst#L200-L216
+  bootPoolFeatures ? [
+    "async_destroy"
+    "bookmarks"
+    "embedded_data"
+    "empty_bpobj"
+    "enabled_txg"
+    "extensible_dataset"
+    "filesystem_limits"
+    "hole_birth"
+    "large_blocks"
+    "lz4_compress"
+    "spacemap_histogram"
+    "userobj_accounting"
+  ]
+
+, # zpool properties
+  bootPoolProperties ? {
+    autoexpand = "on";
+  }
+, # pool-wide filesystem properties
+  bootPoolFilesystemProperties ? {
+    acltype = "posixacl";
+    atime = "off";
+    compression = "on";
+    mountpoint = "legacy";
+    xattr = "sa";
+  }
+, # The name of the ZFS pool
+  rootPoolName ? "tank"
+
+, # zpool properties
+  rootPoolProperties ? {
+    autoexpand = "on";
```

Comment:
it is surprising that this does not work for partitions. Is that behavior noted somewhere so that someone looking at this doesn't waste time trying to make it work?

Reply:
(AFAIK) The reason behind it is zfs, when it owns the whole disk doesn’t
really use a partition: it has a marker partition at the start and a
marker partition at the end of the disk.

In the partition case, zfs is constrained to the partition and is
leaving it up to the user to expand the partition containing zfs, as the
user may choose to not give it to ZFS.

Once the user expands the partition, calling zpool online -e expands the
vdev to use the whole partition. It doesn’t happen automatically
presumably because zed doesn’t get an event saying it’s partition grew,
whereas it can and does get an event saying the whole disk it is on is
now larger.

As far as I know this is all "works as intended" for ZFS. Where would you
expect it to be written down?